### ISSUE TITLE
Slow Gesture/BaseActionItem

### ISSUE THREAD

**Issue by jrieken** [Microsoft/vscode#37917]
* Profile a startup/reload, sort profile by heaviest calls
* The `DomListener`-type is a heavy hitter and takes ~45ms of which ~80% are spend in setting up the `Gesture`

<img width="729" alt="screen shot 2017-11-09 at 11 12 46" src="https://user-images.githubusercontent.com/1794099/32600297-76eba564-c53f-11e7-9dfe-2633851edaa4.png">

[CPU-20171109T111657.cpuprofile.txt](https://github.com/Microsoft/vscode/files/1457227/CPU-20171109T111657.cpuprofile.txt)

It seems that adding the `touch{start/move/end/}` events takes a lot longer than other event listener... Suggestions:

* don't have a listener per `BaseActionItem` but per container (action bar)
* disable gesture when target device does support touch

**Comment by jrieken** [comment]
Not sure who owns this... I started this 5 years ago but I feel responsible for it.

**Comment by bpasero** [comment]
Is there any way to find out if the device is touch enabled or not. @alexandrudima ?

**Comment by isidorn** [comment]
I can look into this.

**Comment by alexandrudima** [comment]
@bpasero I don't know

**Comment by jrieken** [comment]
`'ontouchstart' in window` and some extra checks could do it, see https://stackoverflow.com/questions/14439903/how-can-i-detect-device-touch-support-in-javascript. 

I have tested the about on my MBP and X1 with the results one would expect...

**Comment by roblourens** [comment]
Also https://bugs.chromium.org/p/chromium/issues/detail?id=784027&can=2&q=addeventlistener%20touch&colspec=ID%20Pri%20M%20Stars%20ReleaseBlock%20Component%20Status%20Owner%20Summary%20OS%20Modified from #37818 so it may just get fixed on Chrome's end.

**Comment by isidorn** [comment]
@jrieken I have pushed a potential fix for this, however the results are not that much different than your initial run
Maybe you could get latest and re-run to double check?

I double checked if the touch check works on my mac (false) and x1(true) and all is good.
Also to note is that I made the isTouchDevice a const so in this version that check would be done once on startup instead of making it a func and checking every time.

![screen shot 2017-11-20 at 16 56 31](https://user-images.githubusercontent.com/1926584/33027619-e835c038-ce13-11e7-84a6-ec45a8bb4124.png)

**Comment by jrieken** [comment]
> Maybe you could get latest and re-run to double check?

No, the slow bit is creating `Gesture`-objects and when you just prevent that from one user, all others will still be slow...

**Comment by jrieken** [comment]
There following problems should be tacked (ideally in this order)

1. don't create too many gesture object as its creation is slow, e.g. only create one per action bar not per action bar item. maybe just one for all dom nodes to share?
2. revisit using `Gesture` for click/tap events. That's the main usage and questionable
3. diable gestures for devices without touch

**Comment by isidorn** [comment]
Fixed via https://github.com/Microsoft/vscode/pull/38946
Seems like we save around 40ms on startup.

I verified all the touch behavior works as before including the balistic scroll. Just fyi @joaomoreno @alexandrudima in case somebody reports touch scroll issues in the editor or the tree that you assign to me.

![screen shot 2017-11-22 at 12 55 54](https://user-images.githubusercontent.com/1926584/33126379-908f76f4-cf84-11e7-9af3-94b37922e7f4.png)
